### PR TITLE
CM HeatLearn

### PR DESCRIPTION
**Pull request by gperonato**
Hi @fallahnejad,
Could you do a general review of the CM architecture and features?
@melhk will later do a more thorough code review before the integration in the master branch.

Thank you all!

### CODE REVIEW

**Comment by melhk** [comment]
Hi @gperonato!
Did you see that you have an error in the build-stack?

**Comment by gperonato** [comment]
> Hi @gperonato!
> Did you see that you have an error in the build-stack?

Yes, the build-stack cannot get the API key (which is secret, so not tracked).
This is linked to this: https://github.com/enermaps/enermaps/pull/252

I don't know whether there is a way to implement the continuous integration while preserving the secret key.
Maybe giving access to default data for the test when the API is not accessible?

**Comment by melhk** [comment]
Ok I see. @neodark, do you have an idea?

File: cm/heatlearn/heatlearn.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    v = np.array(list(dic.values()))
+
+    # Get argsort indices
+    sidx = k.argsort()
+
+    # Drop the magic bomb with searchsorted to get the corresponding
+    # places for a in keys (using sorter since a is not necessarily sorted).
+    # Then trace it back to original order with indexing into sidx
+    # Finally index into values for desired output.
+    return v[sidx[np.searchsorted(k, ar, sorter=sidx)]]
```
Comment (melhk) [inline cm/heatlearn/heatlearn.py:66]: What are "ar, k, v, sidx"? Please give an understandable name to your variables.

File: frontend/src/src/settings.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 export const BASE_URL = '/';
-export const INITIAL_MAP_CENTER = [51.505, -0.09];
+export const INITIAL_MAP_CENTER = [46.2, 6.14]; // Geneva
```
Comment (melhk) [inline frontend/src/src/settings.js:2]: Please don't modify the front end that is not directly linked to your CM.

File: cm/heatlearn/heatlearn.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    500: "9216fdce-9855-11eb-b3f9-3e22fb3fc3ab",
+    300: "2dcaa71c-9837-11eb-b3f9-3e22fb3fc3ab",
+}
+
+HDD_MODEL = {"slope": -0.0002855045732455094, "intercept": 1.8116549365250931}
+
+
+def replace_with_dict(ar: np.array, dic: dict = ESM_dict):
+    """Replace numpy values with dict (source: https://stackoverflow.com/a/47171600)."""
+    #
```
Comment (melhk) [inline cm/heatlearn/heatlearn.py:54]: Please remove this.

File: cm/heatlearn/heatlearn.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    tiles = tiles.loc[tiles.intersects(boundary.iloc[0].geometry), :]
+    tiles = tiles.reset_index()
+    if tiles.shape[0] == 0:
+        raise ValueError("No tiles were created.")
+
+    # Prepare raster
+    if len(raster_paths) == 1:
+        raster_path = raster_paths[0]
+    else:
+        raise ValueError("Only a single raster is supported for now.")
```
Comment (melhk) [inline cm/heatlearn/heatlearn.py:206]: Make this check in the worker.py file, before calling the heatlearn function (and check all the other parameters there too).

File: cm/heatlearn/heatlearn.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        """Check whether tile complies with minimum requirements."""
+        with open(
+            os.path.join("models", MODELS[tile_size], "parameters.json"), "r"
+        ) as f:
+            parameters = json.loads(f.read())
+        coverage_ratio = np.sum(matrix == 0) / np.sum(matrix >= 0)
+        if coverage_ratio > parameters["idc_coverage_ratio"]:
+            return True
+        else:
+            return False
```
Comment (melhk) [inline cm/heatlearn/heatlearn.py:231]: Is there a reason why these functions are inside the heatlearn function?

File: cm/heatlearn/worker.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    if not selection["features"]:
+        raise ValueError("The selection must be non-empty.")
+    raster_paths = []
+    for raster in rasters:
+        raster_paths.append(cm_input.get_raster_path(raster))
+    self.validate_params(params)
+
+    tile_size = params["tileSize"]
+    year = params["year"]
+    results = heatlearn(selection, raster_paths, tile_size, year)
```
Comment (melhk) [inline cm/heatlearn/worker.py:32]: Check the parameters before calling heatlearn

**Comment by gperonato** [comment]
FYI, a new branch has been created to be cleaned and rebased to the current master:
https://github.com/enermaps/enermaps/tree/cm_heatlearn_new
However, the functioning of the CM - in line with the wiki and documentation - is better to be seen in this old branch.

I will address the comments here on the new branch.

File: cm/heatlearn/heatlearn.py (hunk 4 above)
Comment (gperonato) [reply cm/heatlearn/heatlearn.py:206]: I think this should stay here. The `else` will take care of merging more rasters (TBD).

**Comment by gperonato** [comment]
Moved here https://github.com/enermaps/enermaps/pull/354

**Pull request closed without merging**